src/main.cpp

```diff
@@ -1488,7 +1489,12 @@ bool OutOfSyncByAge()
     // rules that Bitcoin uses.
     constexpr int64_t maxAge = 90 * 10;
 
-    return GetAdjustedTime() - g_previous_block_time >= maxAge;
+    bool out_of_sync = (GetAdjustedTime() - g_previous_block_time >= maxAge);
+
+    // g_synced_before is an atomic boolean global that keeps track of whether the wallet was ever in sync in this run.
+    if (!out_of_sync) g_synced_before = true;
```

Comment:
Can we move this to `UpdateSyncTime()` in the `!OutOfSyncByAge()` condition? I think it may be confusing to introduce a mutation into a query function, and this one is called all over the place.

Reply:
Actually, using `g_nTimeBestReceived > 0` may eliminate the need for the extra state.

Reply:
Yep. Because the logic to first update g_nTimeBestReceived is identical.